lib/stores/S3Store.js

```diff
@@ -505,23 +505,44 @@ class S3Store extends DataStore {
             this._getMetadata(file_id)
                 .then((metadata) => {
                     return this._retrieveParts(file_id)
-                        .then((parts) => {
-                            return {
-                                parts,
-                                metadata,
-                            };
-                        })
+                        .then((parts) => ({ parts, metadata }))
                         .catch((err) => {
-                            throw err;
+                            // getOffset is called from both the HEAD request and the PATCH request
+                            // (via the _processUpload method). The HEAD request calls it (implicitly)
+                            // with with_parts: false, the _processUpload method calls it (explicitly)
+                            // with with_parts: true. The _processUpload method expects upload parts to be returned,
+                            // which would not work when the upload is completed, since there are no parts anymore.
+                            // TODO: refactor to separate these concerns.
+                            // TODO: write a test for this case.
+                            if (with_parts === true || err.code !== 'NoSuchUpload') {
+                                throw err;
+                            }
+                            return this.client.headObject({ Bucket: this.bucket_name, Key: file_id })
+                                .promise()
+                                .then(() => {
+                                    resolve({
+                                        ...this.cache[file_id].file,
+                                        size: metadata.file.upload_length,
+                                        upload_length: metadata.file.upload_length,
+                                    });
+                                    return { isFinished: true };
```

Comment:
resolve and return mixed seems a bit strange
```suggestion
                                   return {
                                        data: {
                                            ...this.cache[file_id].file,
                                            size: metadata.file.upload_length,
                                            upload_length: metadata.file.upload_length,
                                        },
                                        isFinished: true,
                                    };
```

Reply:
yeah agree.
now it's pretty similar to where we started from :-)
I'll change it

Reply:
ok. that's in now. I am doing more testing now

Reply:
I have finished my testing with the current state and haven't seen any issues. I simulated the error case and I tested the default case.

So from my side this is all good to go.